File: app/src/main/java/com/arcinteractive/spaces/data/model/SpaceMedia.kt
package com.arcinteractive.spaces.data.model

data class SpaceMedia(
    val id: String,
    val spaceId: String? = null,
    val type: MessageType,
    val mediaCategory: String? = null,
    val mediaType: MediaType = inferMediaType(type, mediaCategory),
    val placeholderIconName: String,
    val caption: String?,
    val senderName: String,
    val timestamp: String,
    val mediaStoragePath: String? = null,
    val thumbnailStoragePath: String? = null,
    val mediaNonceBase64: String? = null,
    val thumbnailNonceBase64: String? = null,
    val metadata: EncryptedMediaMetadata? = defaultMetadata(
        id = id,
        mediaType = inferMediaType(type, mediaCategory),
        mediaStoragePath = mediaStoragePath,
        thumbnailStoragePath = thumbnailStoragePath,
        mediaNonceBase64 = mediaNonceBase64,
        thumbnailNonceBase64 = thumbnailNonceBase64
    )
)

private fun inferMediaType(type: MessageType, mediaCategory: String?): MediaType {
    val normalized = mediaCategory?.trim().orEmpty()
    return when {
        normalized.equals("photo", ignoreCase = true) -> MediaType.Photo
        normalized.equals("meme", ignoreCase = true) -> MediaType.Meme
        normalized.equals("gif", ignoreCase = true) -> MediaType.Gif
        normalized.equals("video", ignoreCase = true) -> MediaType.Video
        normalized.equals("file", ignoreCase = true) -> MediaType.File
        normalized.equals("voice", ignoreCase = true) -> MediaType.Voice
        normalized.equals("profilePhoto", ignoreCase = true) -> MediaType.ProfilePhoto
        normalized.equals("coverPhoto", ignoreCase = true) -> MediaType.CoverPhoto
        type == MessageType.Video -> MediaType.Video
        type == MessageType.Meme -> MediaType.Meme
        type == MessageType.Gif -> MediaType.Gif
        type == MessageType.File -> MediaType.File
        else -> MediaType.Photo
    }
}

private fun defaultMetadata(
    id: String,
    mediaType: MediaType,
    mediaStoragePath: String?,
    thumbnailStoragePath: String?,
    mediaNonceBase64: String?,
    thumbnailNonceBase64: String?
): EncryptedMediaMetadata? {
    if (mediaStoragePath.isNullOrBlank() || mediaNonceBase64.isNullOrBlank()) return null
    return EncryptedMediaMetadata(
        mediaId = id,
        mediaType = mediaType,
        storagePath = mediaStoragePath,
        thumbnailStoragePath = thumbnailStoragePath,
        encryptionVersion = "aes-gcm-v1",
        nonce = mediaNonceBase64,
        thumbnailNonce = thumbnailNonceBase64,
        mimeType = "image/jpeg",
        fileSize = 0L,
        uploadedBy = ""
    )
}

File: iOS/Spaces V2/Models/SpaceMedia.swift
import Foundation

struct SpaceMedia: Identifiable, Hashable {
    let id: String
    let spaceID: String?
    let type: MessageType
    let mediaCategory: String?
    let mediaType: MediaType
    let placeholderImageName: String
    let caption: String?
    let senderName: String
    let timestamp: String
    let metadata: EncryptedMediaMetadata?
    let mediaStoragePath: String?
    let thumbnailStoragePath: String?
    let mediaNonceBase64: String?
    let thumbnailNonceBase64: String?
    let galleryItems: [SpaceMedia]?
    let gallerySelectedIndex: Int

    init(
        id: String,
        spaceID: String? = nil,
        type: MessageType,
        mediaCategory: String? = nil,
        mediaType: MediaType? = nil,
        placeholderImageName: String,
        caption: String?,
        senderName: String,
        timestamp: String,
        metadata: EncryptedMediaMetadata? = nil,
        mediaStoragePath: String? = nil,
        thumbnailStoragePath: String? = nil,
        mediaNonceBase64: String? = nil,
        thumbnailNonceBase64: String? = nil,
        galleryItems: [SpaceMedia]? = nil,
        gallerySelectedIndex: Int = 0
    ) {
        self.id = id
        self.spaceID = spaceID
        self.type = type
        self.mediaCategory = mediaCategory
        self.mediaType = mediaType ?? Self.inferredMediaType(from: type, category: mediaCategory)
        self.placeholderImageName = placeholderImageName
        self.caption = caption
        self.senderName = senderName
        self.timestamp = timestamp
        let resolvedMetadata = metadata ?? Self.defaultMetadata(
            id: id,
            mediaType: self.mediaType,
            mediaStoragePath: mediaStoragePath,
            thumbnailStoragePath: thumbnailStoragePath,
            mediaNonceBase64: mediaNonceBase64,
            thumbnailNonceBase64: thumbnailNonceBase64
        )
        self.metadata = resolvedMetadata
        self.mediaStoragePath = resolvedMetadata?.storagePath ?? mediaStoragePath
        self.thumbnailStoragePath = resolvedMetadata?.thumbnailStoragePath ?? thumbnailStoragePath
        self.mediaNonceBase64 = resolvedMetadata?.nonce ?? mediaNonceBase64
        self.thumbnailNonceBase64 = resolvedMetadata?.thumbnailNonce ?? thumbnailNonceBase64
        self.galleryItems = galleryItems
        self.gallerySelectedIndex = gallerySelectedIndex
    }

    init(
        id: UUID,
        spaceID: String? = nil,
        type: MessageType,
        mediaCategory: String? = nil,
        mediaType: MediaType? = nil,
        placeholderImageName: String,
        caption: String?,
        senderName: String,
        timestamp: String,
        metadata: EncryptedMediaMetadata? = nil,
        mediaStoragePath: String? = nil,
        thumbnailStoragePath: String? = nil,
        mediaNonceBase64: String? = nil,
        thumbnailNonceBase64: String? = nil,
        galleryItems: [SpaceMedia]? = nil,
        gallerySelectedIndex: Int = 0
    ) {
        self.init(
            id: id.uuidString,
            spaceID: spaceID,
            type: type,
            mediaCategory: mediaCategory,
            mediaType: mediaType,
            placeholderImageName: placeholderImageName,
            caption: caption,
            senderName: senderName,
            timestamp: timestamp,
            metadata: metadata,
            mediaStoragePath: mediaStoragePath,
            thumbnailStoragePath: thumbnailStoragePath,
            mediaNonceBase64: mediaNonceBase64,
            thumbnailNonceBase64: thumbnailNonceBase64,
            galleryItems: galleryItems,
            gallerySelectedIndex: gallerySelectedIndex
        )
    }

    func withGallery(items: [SpaceMedia], selectedIndex: Int) -> SpaceMedia {
        SpaceMedia(
            id: id,
            spaceID: spaceID,
            type: type,
            mediaCategory: mediaCategory,
            mediaType: mediaType,
            placeholderImageName: placeholderImageName,
            caption: caption,
            senderName: senderName,
            timestamp: timestamp,
            metadata: metadata,
            mediaStoragePath: mediaStoragePath,
            thumbnailStoragePath: thumbnailStoragePath,
            mediaNonceBase64: mediaNonceBase64,
            thumbnailNonceBase64: thumbnailNonceBase64,
            galleryItems: items,
            gallerySelectedIndex: selectedIndex
        )
    }

    private static func inferredMediaType(from type: MessageType, category: String?) -> MediaType {
        if let category, let mediaType = MediaType(rawValue: category) {
            return mediaType
        }

        switch type {
        case .text:
            return .file
        case .image:
            return .photo
        case .video:
            return .video
        case .meme:
            return .meme
        case .gif:
            return .gif
        case .screenshot:
            return .photo
        case .file:
            return .file
        }
    }

    private static func defaultMetadata(
        id: String,
        mediaType: MediaType,
        mediaStoragePath: String?,
        thumbnailStoragePath: String?,
        mediaNonceBase64: String?,
        thumbnailNonceBase64: String?
    ) -> EncryptedMediaMetadata? {
        guard let mediaStoragePath, let mediaNonceBase64 else { return nil }
        return EncryptedMediaMetadata(
            mediaId: id,
            mediaType: mediaType,
            storagePath: mediaStoragePath,
            thumbnailStoragePath: thumbnailStoragePath,
            encryptionVersion: "aes-gcm-v1",
            nonce: mediaNonceBase64,
            thumbnailNonce: thumbnailNonceBase64,
            mimeType: "image/jpeg",
            fileSize: 0,
            width: nil,
            height: nil,
            duration: nil,
            createdAt: nil,
            uploadedBy: ""
        )
    }
}
